### PR TITLE
Updates for GEOS-Chem

### PR DESCRIPTION
**Pull request by lizziel**
### Name and Institution (Required)

Name: Lizzie Lundgren
Institution: Harvard University

### Describe the update

This PR brings into the main branch all remaining changes made during development of Cloud-J for use in GEOS-Chem. This includes changes for using GEOS-Chem with Cloud-J photolysis in multiple configurations: GEOS-Chem Classic, GCHP, GEOS, and CESM. It also includes several bug fixes for compatibility with GNU and Intel compilers, and one Cloud-J standalone bug fix for computing ice water path input to Cloud_JX. This PR also includes a change in expected spacing of entries in input file `FJX_scat-aer.dat` which will affect Cloud-J standalone users who have custom LUTs outside of the ones stored in this repository.

### Expected changes

See the [Changelog](https://github.com/geoschem/Cloud-J/pull/2/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed) (new file added in this PR) for details of all changes in Cloud-J.

### Reference(s)

None

Tagging @pratherUCI for review

### CODE REVIEW

**Comment by lizziel** [comment]
Merging this in is on the back-burner because it will break compatibility with the Cloud-J standalone tables files.

**Comment by lizziel** [comment]
This PR is almost ready for review. I pushed the latest updates for GEOS-Chem in offline, GEOS, and CESM and now need to address the resulting conflicts in the standalone model.

**Comment by lizziel** [comment]
This PR is now ready for review. I confirmed that the changes cause no differences in Cloud-J standalone output when using default settings.

File: src/Core/cldj_error_mod.F90
Comment (yantosca) [inline src/Core/cldj_error_mod.F90:36]: Just a generic note: In a future version we should think about propagating a success/failure error code up to the main program level.  If we stop with an error when running an instance of GEOS-Chem in an MPI environment, we could hang a simulation (because one core would be stopped but the others would not).  Something to think about for down the road...

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:141]: NOTE: You don't need to have multiple `implicit none` statements.  One at the top of the module is sufficient.

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:220]: For safety's sake, consider also initializing these output variables at the top of the routine
```F90
SWMSQ = 0.0d0
OD18 = 0.0d0
LDARK = .FALSE.
```

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:1896]: For safety's sake, consider setting these output arrays to 0 at the start of the routine:
```F90
DDENS = 0.0d0
QQEXT = 0.0d0
SSALB = 0.0d0
SSLEG = 0.0d0
```

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:219]: When assigning arrays, using `(:)`, `(:,:)`, etc. is not needed.  It also may be more efficient to omit these.
```F90
VALJXX = 0.0d0
FFXTAU = 0.0d0
```

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:279]: This can be `TAUG_LLNL = 0.0d0`

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:304]: This can be `SLEG = 0.0d0`

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:1940]: For safety's sake, consider zeroing the output variables at the top of the routine
```F90
DDENS = 0.0d0
QQEXT = 0.0d0
SSALB = 0.0d0
SSLEG = 0.0d0
```

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2479]: This can be `PATH1 = 0.0d0`

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2457]: These can be 
```F90
ZANG1 = 0d0
...
PATH1 = 0d0
```

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2506]: This can be `ZANG1 = 0.0d0`, `PATH1= 0.0d0`

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2556]: See similar comment above

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2459]: This can be `PATH1 = 0.0d0`

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2800]: For safety's sake, consider setting the output variable `JXTRA = 0` at the top of the routine.

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:1985]: See similar comment above about setting output variables to zero

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2026]: See similar comment about setting output variables to zero at top of routine

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2069]: See similar comment about setting output variables to zero at top of routine

File: src/Core/cldj_fjx_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_fjx_sub_mod.F90:2121]: See similar comment about setting output variables to 0 at top of routine

File: src/Core/cldj_osa_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_osa_sub_mod.F90:25]: @lizziel @pratherUCI: Am wondering if these parameters should be `REAL*8` and if numerical constants should use `d` exponents.  Otherwise we would potentially incur numerical roundoff.  I would leave it as-is for now but maybe we can investigate more in the future.

File: src/Core/cldj_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_sub_mod.F90:138]: See code immediately below this.  The loop over L isn't necessary to initialize the arrays.  The code could be:
```F90
      LPRTJ0  = LPRTJ
      JCOUNT  = 0
      NICA    = 0
      LWPX    = 0.d0
      IWPX    = 0.d0
      REFFLX  = 0.d0
      REFFIX  = 0.d0
      VALJXX  = 0.d0    ! zero J's Heating R's in case LDARK is returned
      SKPERD  = 0.d0
      SWMSQ   = 0.d0
      VALJXXX = 0.d0   ! zero the PHOTOJ equivalents for wtd averaging
      SKPERDD = 0.d0
      SWMSQQ  = 0.d0
      OD18    = 0.d0
      OD18Q   = 0.d0
```

File: src/Core/cldj_sub_mod.F90
Comment (yantosca) [inline src/Core/cldj_sub_mod.F90:205]: This loop isn't necessary to zero the variables.  We can just use:
```F90
LWPX   = 0.d0
IWPX   = 0.d0
REFFLX = 0.d0
REFFIX = 0.d0
```

**Review by yantosca — CHANGES_REQUESTED** [review]
Hi @lizziel, thanks for this.  I made several minor comments about mostly formatting, precision, and array initialization issues.  Let me know if you have any more questions.

**Comment by lizziel** [comment]
Thanks @yantosca for your detailed review! I tried to keep the Cloud-J as close to the original as possible to avoid divergence. Your comments will be very helpful when we go through and do general improvements to the code. 

@pratherUCI, once we start making general code improvements it will be more difficult to bring in updates you have based on older versions. Would you be able to start working off of this version (once merged) moving forward to avoid problems down the road?

**Review by yantosca — APPROVED** [review]
Thanks @lizziel.  I agree, we can make no-diff updates later.  No need to hold up this PR.  Good to merge!

**Comment by pratherUCI** [comment]
I went thru all the changes.  They make sense and indeed help.  I approve.
Michael

There is no reason why the standalone version that  I continue to develop could not now begin with the new version.

There are two remaining issues for us:
1) The next version of Cloud-J standalone is v8.0c, it has been changed (minorly) to include H2O absorption.  At the time I also went through all the old commons and removed dead variables (as you did).  I will check my list.  So how is best to proceed. 
BTW, this code enabled me to include NO2 absorption (small effects vs H2O).    The major change is in one input table, a few lines of code, and then the clean up.  This is not urgent, the only change is the H2O abs.
Prather, Michael (2023). An updated cloud-overlap photolysis module for atmospheric chemistry models, UCI Cloud-J v8.0, with near-UV H2O absorption [Dataset]. Dryad. https://doi.org/10.7280/D1Q398

2) I have intentionally ignored development for the aerosol code until there was a reason to update.  GEOS Chem should have updated aerosol tables that address rel hum, etc.  We need someone who is interested in the aerosol simulations and I can work with them.

Michael

**Comment by lizziel** [comment]
Hi @pratherUCI, thanks for reviewing the code. I will create a new GitHub issue to discuss putting in the next version of Cloud-J.

**Pull request merged**
